File: findAllAnagramsInString/findAllAnagramsInAString.cpp
/**
 *  https://leetcode.com/problems/find-all-anagrams-in-a-string/description/
 *
 *  Runtime: O(s*p) where p is the size of the substring p and s is the size of the string s 
*/

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> solution;
        int lenP = p.size();
        
        if (lenP > s.size()) {
            return solution;
        }
        for (int i = 0; i < s.size()-lenP+1 ; i++){
            // if the previous i was an anagram...
            if ( i > 0 && solution.size() > 0 ) {
                if (solution[solution.size()-1] == i-1) {
                    // Check if the next char is the same as the previous
                    if ( i+lenP <= s.size()-1){
                        if ( s[i-1] == s[i+lenP-1] ) { 
                            // Same chars implies that it is also an anagram.
                            solution.push_back(i);
                            continue;
                        } 
                    }
                }
            } 
            
            if (isAnagram (s.substr(i,lenP), p)) {
                solution.push_back(i);
            }
        }
        return solution;
    }
    
    bool isAnagram (string a, string b) {
        if (a.size() != b.size()) {
            return false;
        }
        int letterCount[26] = {0};
        
        // Count the letters in first string
        for (int i = 0; i < a.size(); i++) {
            char ch = a[i];
            if (letterCount[ch-'a'] > 0) {
                letterCount[ch-'a']++;
            } else {
                letterCount[ch-'a'] = 1;
            }
        }
        
        // Reduce the counts as we go through the second string
        for ( int i = 0; i < b.size(); i++) {
            char ch = b[i];
            int count = letterCount[ch-'a'];
            if (count == 0 ){
                return false;
            } else if ( count > 0) {
                letterCount[ch-'a']--; 
            }
        }
        
        return true;
    }
};

File: findAllAnagramsInString/findAllAnagramsInAString.java
/**
 *  https://leetcode.com/problems/find-all-anagrams-in-a-string/description/ *
 *  Runtime: O(n lg(n)) due to using a hash map. Switching from hash map
 *  to a character array of size 256 ( 1 byte ) would make each put from 
 *  the hash map from O(lg(n)) to O(1).
*/

class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> solution = new LinkedList<>();
        
        if ( p.length() > s.length() ) {
            return solution;
        }
        
        int begin = 0, end = 0, counter = 0;
        Map<Character, Integer> charMap = new HashMap<>();
        
        // Count the characters in the target string
        for ( char c : p.toCharArray() ) {
            charMap.put (c, charMap.getOrDefault(c,0) + 1);
        }
        
        counter = charMap.size();
        
        while ( end < s.length()) {
            char endChar = s.charAt(end);
            if( charMap.containsKey(endChar) ) {
                charMap.put(endChar, charMap.get(endChar)-1);
                
                // If we we have enough of the character
                if ( charMap.get(endChar) == 0 ) {
                    // Decrease the counter to say that there is a match.
                    counter--;
                }
            }
            end++;
            
            // While the counter is 0 (because if there is a bunch
            // of characters in the beginning that aren't part of the anagram
            // we can skip them)
            while ( counter == 0 ) {
                char beginChar = s.charAt(begin);
                if ( charMap.containsKey(beginChar) ) {
                    charMap.put(beginChar, charMap.get(beginChar)+1);
                    if(charMap.get(beginChar) > 0){
                        counter++;
                    }
                }
                if ( end-begin == p.length() ) {
                    solution.add(begin);
                }
                begin++;
            }
        }
        return solution;
    }
}